### ISSUE TITLE
Bulk INSERT a Pandas dataframe

### ISSUE THREAD

**Issue by felipe0216** [sqlalchemy/sqlalchemy#5009]
Hello!

I am currently working with a timeseries pandas dataframe (I will call it Cashflow) that contains 200+ columns and 10,000+ rows, I was wondering what is the appropriate way to INSERT this "Cashflow" dataframe into a table in the database. The database table  already exists and it has already been reflected, so I have all the columns and the schema for it (so the class is created).

I am trying to work out the logic for this. I have seen that the "Session.bulk_" methods use dictionaries, so I am thinking on converting the Cashflow dataframe into a dictionary, but then I can't think of a way to iterate over the 200+ columns and 10,000+ rows of the dataframe to create 10,000 instances of the class "Cashflow" and then insert them all at once with the "bulk" method. Also, dictionaries are an unordered array and since this is a timeseries data, I am not sure about it.

 I am trying to workout a way using ORM, but if you can think of a way using CORE, I would be more than happy as well.

UPDATE: The dataframe is now 420,000 rows and 57 columns

**Comment by gordthompson** [comment]
Since you already have the data in a pandas DataFrame, why can't you just use [to_sql(https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_sql.html)?

**Comment by felipe0216** [comment]
How would that look like for a dataframe that big? I mean, I can imagine the insert being something like this:
("INSERT INTO cashflow VALUES ( ? , ? , ? , ? ........)",(value1, value2, ......)), but how would you do it for the 200+ columns? 

Also, wouldn't it be slower than SQLAlchemy?

**Comment by gordthompson** [comment]
>  I can imagine the insert being something like this:

The actual form of the INSERT statement(s) would depend on the database you are using and the options you specify for `to_sql`. 

> Also, wouldn't it be slower than SQLAlchemy?

If you use `to_sql(df, engine, ...)` then pandas will be using SQLAlchemy on your behalf. The question then becomes whether you can come up with your own code that is more efficient than the `to_sql` code in pandas that has been around for quite a while and has been "field tested" by many, many users.

As for the size of your DataFrame, 200 * 10_000 is only 2_000_000 elements, which is significant but not outrageous. What does `sys.getsizeof(your_df)` return?

**Comment by felipe0216** [comment]
Hello @gordthompson ,

Thanks for your answer, it was very helpful. There result of my `sys.getsizeof(your_df)` is 6771509. it is a dataframe 5000 rows and 57 columns long

I have made a few experiment running the ORM, Core, and the classic DBAPI inserting methods to see which one is the fastest and oddly enough, the DBAPI "Executemany" code run faster than any of the SQLAlchemy options. Do you know why that could have been? (I have attached the code below)
```

def init_sqlalchemy():
    global engine
    Session.configure(bind=engine, autoflush=False, expire_on_commit=False)
    Base.metadata.drop_all(engine)
    Base.metadata.create_all(engine)


def sqlalchemy_orm_bulk_insert(class_object, table, n=100000):
    init_sqlalchemy()
    t0 = time.time()
    session.bulk_insert_mappings(class_object,
                                 table[:n].to_dict(orient="record"))

    session.commit()
    print("SQLAlchemy ORM bulk_insert_mappings(): Total time for " + str(n) +
          " records " + str(time.time() - t0) + " secs")


def sqlalchemy_core(class_object, table, n=100000):
    init_sqlalchemy()
    t0 = time.time()
    engine.execute(class_object.__table__.insert(),
                   table[:n].to_dict(orient="record"))
    print("SQLAlchemy Core: Total time for " + str(n) + " records " +
          str(time.time() - t0) + " secs")

def dbapi_insert_executemany_itertuple(db_table, df, n=100000):
    import pyodbc 
    import time
    cnxn = pyodbc.connect('Driver={SQL Server};'\
                                "SERVER=BGFSQPR03;"\
                                "Database=BGT;"\
                                "UID=Felipe.Araya;"\
                                "PWD=Be1m0nt$$1909;"\
                                "DSN=TreasuryMSSQL;"
                                "Trusted_Connection=yes")
    t0 = time.time()
    cols = ",".join([str(i) for i in df.columns.tolist()])
    
    cursor = cnxn.cursor()
    cursor.fast_executemany = True
    cursor.executemany("INSERT INTO "+ str(db_table) +"( " + cols + " ) VALUES (" + "?,"*(len(df.columns)-1) + "?)", df[:n].itertuples(index=False, name=None))
    cursor.commit()
    cnxn.close()

    print("DBAPI Executemany-Itertuple insertion: Total time for: " + str(n) + " records " + str(time.time()-t0) + " secs to insert")


def sqlalchemy_insert_to_sql_(db_table, df, n=100000):
    import time
    global engine
    t0 = time.time()
    df[:n].to_sql(str(db_table), con = engine, schema="dbo", if_exists = 'append', index= False)
    print("SQLAclhemy - to_sql insertion: Total time for: " + str(n) + " records " + str(time.time()-t0) + " secs to insert")

sqlalchemy_orm_bulk_insert(AssetCashFlows, cashflow_extended, n=5000)
sqlalchemy_core(AssetCashFlows,cashflow_extended, n= 5000)
dbapi_insert_executemany_itertuple("dbo.AssetCashFlows", cashflow_extended, n=5000)
sqlalchemy_insert_to_sql_("AssetCashFlows",cashflow_extended, n=5000)
```
Outputs:
66 secs
65 secs
63 secs
65 secs

**Comment by gordthompson** [comment]
They're all pretty close. 65 seconds is just 3% more than 63 seconds. Have you repeated the tests enough times (ideally changing up the order of the individual tests) to confirm that the 3% difference is reasonably consistent?

**Comment by felipe0216** [comment]
Hello @gordthompson,

Yes, I have repeated the test multiple times with more and less rows and the dbapi_executemany method always works a bit faster than any of the other ones. Is there anything wrong with the code? I really don't understand why this could be happening.

PD: it is not always a 3% faster, but it is always a few seconds faster regardless of the amount of rows.

**Comment by gordthompson** [comment]
> it is always a few seconds faster regardless of the amount of rows

So it sounds like those few seconds are the overhead of establishing the SQLAlchemy environment and getting pandas to work with it. I did a little test to compare a raw pyodbc connection vs. pandas `to_sql` using SQL Server. SQL Profiler showed that ultimately both tests ended up doing

```sql
exec sp_prepare @p1 output,N'@P1 int,@P2 nvarchar(50)',N'INSERT INTO issue5009 (id, txt) VALUES (@P1, @P2)',1
exec sp_execute 1,1,N'foo'
exec sp_execute 1,2,N'bar'
exec sp_unprepare 1
```

but while the vanilla pyodbc approach did

```sql
exec [sys].sp_datatype_info_100 12,@ODBCVer=4
exec [sys].sp_datatype_info_100 -9,@ODBCVer=4
exec [sys].sp_datatype_info_100 -3,@ODBCVer=4
exec [sys].sp_datatype_info_100 93,@ODBCVer=4
DELETE FROM issue5009
 set fmtonly on select id,txt from issue5009 where 1=2 set fmtonly off
exec sp_prepare @p1 output,N'@P1 int,@P2 nvarchar(50)',N'INSERT INTO issue5009 (id, txt) VALUES (@P1, @P2)',1
exec sp_execute 1,1,N'foo'
exec sp_execute 1,2,N'bar'
exec sp_unprepare 1
```

the pandas `to_sql` approach did

```sql
exec [sys].sp_datatype_info_100 12,@ODBCVer=4
exec [sys].sp_datatype_info_100 -9,@ODBCVer=4
exec [sys].sp_datatype_info_100 -3,@ODBCVer=4
exec [sys].sp_datatype_info_100 93,@ODBCVer=4
SELECT CAST(SERVERPROPERTY('ProductVersion') AS VARCHAR)
SELECT schema_name()

                  SELECT CASE transaction_isolation_level
                    WHEN 0 THEN NULL
                    WHEN 1 THEN 'READ UNCOMMITTED'
                    WHEN 2 THEN 'READ COMMITTED'
                    WHEN 3 THEN 'REPEATABLE READ'
                    WHEN 4 THEN 'SERIALIZABLE'
                    WHEN 5 THEN 'SNAPSHOT' END AS TRANSACTION_ISOLATION_LEVEL
                    FROM sys.dm_exec_sessions
                    where session_id = @@SPID
                  
SELECT CAST('test plain returns' AS VARCHAR(60)) AS ['test plain returns']
SELECT CAST('test unicode returns' AS NVARCHAR(60)) AS ['test unicode returns']
DELETE FROM issue5009
 set fmtonly on select id,txt from issue5009 where 1=2 set fmtonly off
exec sp_prepare @p1 output,N'@P1 int,@P2 nvarchar(50)',N'INSERT INTO issue5009 (id, txt) VALUES (@P1, @P2)',1
exec sp_execute 1,1,N'foo'
exec sp_execute 1,2,N'bar'
exec sp_unprepare 1
```

Frankly, I'd be surprised if just those extra five (5) SELECTs consistently took "a few seconds" to run, but then SQLAlchemy/pandas is presumably using the *results* of those queries to do clever things. That would add to the fixed cost of the batch, but those extra few moments would become an ever-smaller percentage of the total cost as the size of the batch increases.

TL;DR: If you're doing lots of small uploads in a tight loop then the savings accrued by using a raw DBAPI method might be significant. Otherwise, don't worry about it and just use the approach that suits you best.

**Comment by felipe0216** [comment]
Thanks for your answer! Even though I am still a bit baffled about the results since I thought that SQLAlchemy was mean to be always faster than the normal DBAPI calls, your answer helps me understand a bit better. Thanks!
